segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/extendeds3/ExtendedS3Storage.java

```diff
@@ -0,0 +1,445 @@
+
+        if (!client.listObjects(config.getBucket(), config.getRoot() + streamSegmentName).getObjects().isEmpty()) {
+            throw new StreamSegmentExistsException(streamSegmentName);
+        }
+
+        S3ObjectMetadata metadata = new S3ObjectMetadata();
+        metadata.setContentLength((long) 0);
+
+        PutObjectRequest request = new PutObjectRequest(config.getBucket(),
+                config.getRoot() + streamSegmentName,
+                (Object) null);
+
+        AccessControlList acl = new AccessControlList();
+        acl.addGrants(new Grant[]{
+                new Grant(new CanonicalUser(config.getAccessKey(), config.getAccessKey()),
+                        Permission.FULL_CONTROL)
+        });
+
+        request.setAcl(acl);
+
+        client.putObject(request);
+
+        LoggerHelpers.traceLeave(log, "create", traceId);
+        return syncGetStreamSegmentInfo(streamSegmentName);
+    }
+
+    private Void syncWrite(SegmentHandle handle, long offset, InputStream data, int length) throws StreamSegmentSealedException, BadOffsetException {
+        Preconditions.checkArgument(!handle.isReadOnly(), "handle must not be read-only.");
+
+        long traceId = LoggerHelpers.traceEnter(log, "write", handle.getSegmentName(), offset, length);
+
+        SegmentProperties si = syncGetStreamSegmentInfo(handle.getSegmentName());
+
+        if (si.isSealed()) {
+            throw new StreamSegmentSealedException(handle.getSegmentName());
+        }
+
+        if (si.getLength() != offset) {
+            throw new BadOffsetException(handle.getSegmentName(), si.getLength(), offset);
+        }
+
+        client.putObject(this.config.getBucket(), this.config.getRoot() + handle.getSegmentName(),
+                Range.fromOffsetLength(offset, length), data);
+        LoggerHelpers.traceLeave(log, "write", traceId);
+        return null;
+    }
+
+    private Void syncSeal(SegmentHandle handle) {
+        Preconditions.checkArgument(!handle.isReadOnly(), "handle must not be read-only.");
+
+        long traceId = LoggerHelpers.traceEnter(log, "seal", handle.getSegmentName());
+
+        AccessControlList acl = client.getObjectAcl(config.getBucket(),
+                config.getRoot() + handle.getSegmentName());
+        acl.getGrants().clear();
+        acl.addGrants(new Grant[]{new Grant(new CanonicalUser(config.getAccessKey(), config.getAccessKey()),
+                Permission.READ)});
+
+        client.setObjectAcl(
+                new SetObjectAclRequest(config.getBucket(), config.getRoot() + handle.getSegmentName()).withAcl(acl));
+        LoggerHelpers.traceLeave(log, "seal", traceId);
+        return null;
+    }
+
+    /**
+     * The concat is implemented using extended S3 implementation of multipart copy API. Please see here for more details on multipart copy.
+     * http://docs.aws.amazon.com/AmazonS3/latest/dev/CopyingObjctsUsingLLJavaMPUapi.html
```

Comment:
The URL reference is good, but the description here is not clarifying much. I think what's happening is the following.

To concatenate, we are copying a target segment T and a source segment S to T, so essentially we are doing T <- T + S. This multipart copy apparently guarantees atomicity, in which case T is left as it was originally in the case `completeMultipartUpload` is not invoked. It is unclear to me what happens in the case `completeMultipartUpload` is never invoked, e.g., due to a crash. Is it the case that a temporary object remains forever in the system?

Please improve this description and state the precise guarantees.